crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1442 @@
+            .and_then(|quilt_index| quilt_index.get_quilt_patch_by_identifier(identifier))
+            .and_then(|quilt_patch| self.get_blob_by_quilt_patch(quilt_patch))
+    }
+
+    fn add_slivers(&mut self, slivers: &'a [&'a SliverData<Secondary>]) {
+        for sliver in slivers {
+            self.slivers.insert(sliver.index, sliver);
+        }
+    }
+}
+
+impl<'a> QuiltDecoderV1<'a> {
+    /// Creates a new QuiltDecoderV1 with the given slivers.
+    pub fn new(slivers: &'a [&'a SliverData<Secondary>]) -> Self {
+        Self {
+            slivers: slivers
+                .iter()
+                .map(|s| (s.index, *s))
+                .collect::<HashMap<_, _>>(),
+            quilt_index: None,
+        }
+    }
+
+    /// Creates a new QuiltDecoderV1 with the given slivers, and a quilt index.
+    pub fn new_with_quilt_index(
+        slivers: &'a [&'a SliverData<Secondary>],
+        quilt_index: QuiltIndexV1,
+    ) -> Self {
+        Self {
+            slivers: slivers.iter().map(|s| (s.index, *s)).collect(),
+            quilt_index: Some(quilt_index),
+        }
+    }
+
+    /// Get the blob represented by the quilt patch.
+    fn get_blob_by_quilt_patch(&self, quilt_patch: &QuiltPatchV1) -> Result<Vec<u8>, QuiltError> {
+        let start_idx = usize::from(quilt_patch.start_index);
+        let end_idx = usize::from(quilt_patch.end_index);
+
+        self.check_missing_slivers(start_idx, end_idx)?;
+
+        let unencoded_length = usize::try_from(quilt_patch.unencoded_length())
+            .expect("unencoded_length should fit in usize");
+
+        // Extract and reconstruct the blob.
+        let mut blob = Vec::with_capacity(unencoded_length);
+
+        // Collect data from the appropriate slivers.
+        for i in start_idx..end_idx {
+            let sliver_idx = SliverIndex(i as u16);
+            if let Some(sliver) = self.slivers.get(&sliver_idx) {
+                let remaining_needed = unencoded_length - blob.len();
+                blob.extend_from_slice(
+                    &sliver.symbols.data()[..remaining_needed.min(sliver.symbols.data().len())],
+                );
+            } else {
+                return Err(QuiltError::Other(format!(
+                    "Unexpected state, sliver {} not found",
+                    sliver_idx
+                )));
+            }
+        }
+
+        Ok(blob)
+    }
+
+    /// Checks if the desired slivers are missing.
+    fn check_missing_slivers(&self, start_idx: usize, end_idx: usize) -> Result<(), QuiltError> {
+        let mut missing_slivers = Vec::new();
+        for i in start_idx..end_idx {
+            let sliver_idx = SliverIndex(i as u16);
+            if !self.slivers.contains_key(&sliver_idx) {
+                missing_slivers.push(sliver_idx);
+            }
+        }
+        if !missing_slivers.is_empty() {
+            return Err(QuiltError::MissingSlivers(missing_slivers));
+        }
+        Ok(())
+    }
+}
+
+mod utils {
```

Comment:
Remark (note to self): Still need to review utils and tests in detail.

Reply:
I've reviewed the utils now in detail, but only did a quick pass over the tests.